src/connectionpools.jl

```diff
@@ -0,0 +1,259 @@
+
+The `idle` field is a timestamp to track when a `Connection` was returned to a `Pod` and became idle.
+
+The `count` field keeps track of how many times the connection has been used.
+"""
+mutable struct ConnectionTracker{T}
+    conn::T
+    idle::Float64
+    count::Int
+end
+
+ConnectionTracker(conn::T) where {T} = ConnectionTracker(conn, time(), 1)
+
+"""
+    Pod(T; max::Int, idle::Int, reuse::Int)
+
+A threadsafe object for managing a pool of and the reuse of `Connection` objects (see [`ConnectionTracker`](@ref)).
+
+A Pod manages a collection of `Connection`s and the following keyword arguments allow configuring the management thereof:
+
+  * `max::Int=typemax(Int)`: controls the max # of currently acquired `Connection`s allowed
+  * `idle::Int=typemax(Int)`: controls the max # of seconds a `Connection` may be idle before it should be closed and not reused
+  * `reuse::Int=typemax(Int)`: controls the max # of times a `Connection` may be reused before it should be closed
+
+After creating a `Pod`, `Connection`s can be acquired by calling [`acquire`](@ref) and MUST
+be subsequently released by calling [`release`](@ref).
+"""
+struct Pod{T}
+    # this lock/condition protects the `conns` Vector and `active` Dict
+    # no changes to either field should be made without holding this lock
+    lock::Threads.Condition
+    conns::Vector{ConnectionTracker{T}}
+    active::Dict{Any, ConnectionTracker{T}}
+    max::Int
+    idle::Int
+    reuse::Int
+end
+
+const MAX = typemax(Int)
+
+function Pod(T; max::Int=MAX, idle::Int=MAX, reuse::Int=MAX)
+    return Pod(Threads.Condition(), ConnectionTracker{T}[], Dict{Any, ConnectionTracker{T}}(), max, idle, reuse)
+end
+
+# check if an idle `Connection` is still valid to be reused
+function isvalid(pod::Pod{C}, conn::ConnectionTracker{C}) where {C}
+    if (time() - conn.idle) > pod.idle
+        # println("connection idle timeout")
+        # if the connection has been idle too long, close it
+        close(conn.conn)
+    elseif conn.count >= pod.reuse
+        # println("connection over reuse limit")
+        # if the connection has been used too many times, close it
+        close(conn.conn)
+    elseif isopen(conn.conn)
+        # println("found a valid connection to reuse")
+        # dump(conn.conn)
+        # otherwise, if the connection is open, this is a valid connection we can use!
+        return true
+    else
+        # println("connection no longer open")
+    end
+    return false
+end
+
+"""
+    acquire(f, pod::Pod{C}) -> C
+
+Check first for existing `Connection`s in a `Pod` still valid to reuse,
+and if so, return one. If no existing `Connection` is available for reuse,
+call the provided function `f()`, which must return a new connection instance of type `C`.
+This new connection instance will be tracked by the `Pod` and MUST be returned to the `Pod`
+after use by calling `release(pod, conn)`.
+"""
+function acquire(f, pod::Pod)
+    lock(pod.lock)
+    try
```

Comment:
I think it may be simpler to restructure the `acquire` loop as follows, where we always use the buffer to pass connections rather than passing them as the return value of `wait()`.  The difficulty with wait/notify is that you've got to be very careful synchronizing both sides and making sure there's tasks ready to pick up the connection as the return of `wait`... it seems subtle and difficult to get correct.

OTOH using the buffer we always know where our connections are stored. We suffer slightly more memory operations, but I doubt those are a problem compared to the costs of task switching etc.

```julia
lock(pod.lock)
try
    while true
        # We hold the lock and no other task can interfere with the pod.
        # Check our caches
        while !isempty(pod.conns)
              # Easy, just remove one while holding the lock.
        end
        if length(pod.active) < pod.max
             # Asyncronously call `f()` to open a connection and nodify(pod.lock) with it
        end
        wait(pod.lock)
        # At this point, we should have the available released or newly open connection in
        # `pod.conns` so just loop back to the start. We now hold the lock again so no
        # other tasks should be able to intervene and steal the available connection from us.
    end
end
```

To go with this, release would need to pass returned connections via `pod.conns` rather than via an argument of `notify()`.

Reply:
I actually originally structured the code this way, but found it simpler to use the `wait`/`notify` mechanism to directly "pass" the released connection to a waiting `acquire` task.

Reply:
Ok.  Would have been interesting to see the original, because I felt the structure of the wait loop was still a bit confusing and not obviously correct. Ah well, efficient and correct concurrent programming is (currently) still hard in Julia.